### PR TITLE
Remain on the edit view when saving a snippet draft

### LINKED ISSUE #9725: Saving a draft snippet should remain on the editor instead of redirecting to the listing page
When saving a snippet with a `DraftStateMixin` applied, Wagtail responds with a redirect to the listing page. This is fine (and is the existing behaviour) for snippets without `DraftStateMixin` (or when doing the Publish action instead), as the changes are made to the "live" instance immediately.

However, when saving a draft, it's desirable to remain on the editor as users may want to make additional changes. This is even more so when saving a publishing schedule, as it also uses the save draft mechanism. By redirecting to the listing view, users have to reopen the editor before they can actually schedule the publishing.

Thus, to improve the user experience (and to match the behaviour of the page editor), Wagtail's response should load the editor instead of a redirect to the listing page.

### PR DESCRIPTION
**Pull request by laymonage**
Fixes #9725, requires #9709 ~~and #9711~~.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Comment by laymonage** [comment]
No idea why the permission tests that rely on the `post_migrate` signal handler fail on CI... it works fine on my local machine 🤔

File: wagtail/admin/views/generic/models.py [hunk 1]
```diff
@@ -398,6 +398,11 @@ def get_add_url(self):
             )
         return reverse(self.add_url_name)
 
+    def get_edit_url(self):
+        if self.edit_url_name is None:
+            return None
```
Comment (gasman) [inline wagtail/admin/views/generic/models.py:403]: As far as I can see we don't have any meaningful fallback behaviour in place for when this returns None, so it probably makes sense for this to raise an ImproperlyConfigured exception like the other `get_*_url` methods.

While we're at it, `get_success_buttons` could also be updated to call `get_edit_url`.

**Review by gasman — CHANGES_REQUESTED** [review]
Just a small suggested tweak to `get_edit_url` - other than that, all looks good to me.

File: wagtail/admin/views/generic/models.py (hunk 1 above)
Comment (laymonage) [reply wagtail/admin/views/generic/models.py]: Ah, you're right. I was hesitant to add the fallback as I thought the generic view itself doesn't really use the `edit_url_name`, but now I see that the logic is already in `get_success_buttons`.

Done, thanks!

**Review by gasman — APPROVED** [review]

**Pull request merged**